### PR TITLE
[IFT] Implement support for parsing and applying Table Keyed patches

### PR DESCRIPTION
**Pull request by garretrieger**
Specification: https://w3c.github.io/IFT/Overview.html#table-keyed

### CODE REVIEW

**Review by cmyr — APPROVED** [review]
Okay a bunch of little observations mostly around the API, but overall this looks reasonable to me.

File: incremental-font-transfer/src/font_patch.rs [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+        &self,
+        compatibility_id: &[u32; 4],
+        patch_data: &[u8],
+        encoding: PatchEncoding,
+    ) -> Result<Vec<u8>, ReadError> {
+        // TODO(garretrieger): Return a custom error type instead of ReadError?
+        let patch = match encoding {
+            PatchEncoding::TableKeyed { .. } => parse_table_keyed_patch(patch_data)?,
+            PatchEncoding::GlyphKeyed => {
+                todo!()
```
[the remark this answers is not in the document]
Comment (garretrieger) [reply incremental-font-transfer/src/font_patch.rs:59]: Yeah, I'll be implementing this next so panic for now is fine.

File: incremental-font-transfer/src/font_patch.rs [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+    ///
+    /// Applies the patch to this base. In the base the patch is associated with the supplied
+    /// compatibility_id and has the specified encoding.
+    ///
+    /// Returns the byte data for the new font produced as a result of the patch application.
+    fn apply_patch(
+        &self,
+        compatibility_id: &[u32; 4],
+        patch_data: &[u8],
+        encoding: PatchEncoding,
```
[the remark this answers is not in the document]
Comment (garretrieger) [reply incremental-font-transfer/src/font_patch.rs:35]: That's right. I think I have a good idea how to rework this in conjuction with the patchmap API, will give that a try.

File: incremental-font-transfer/src/font_patch.rs [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+            table_patch.max_uncompressed_length() as usize,
+        ),
+        (None, false) => return Err(ReadError::ValidationError),
+        _ => shared_brotli_decode(stream, None, table_patch.max_uncompressed_length() as usize),
+    };
+
+    r.map_err(|decode_error| match decode_error {
+        DecodeError::InitFailure => ReadError::MalformedData("Failure to init brotli encoder."),
+        DecodeError::InvalidStream => ReadError::MalformedData("Malformed brotli stream."),
+        DecodeError::InvalidDictionary => ReadError::MalformedData("Malformed dictionary."),
```
[the remark this answers is not in the document]
Comment (garretrieger) [reply incremental-font-transfer/src/font_patch.rs:186]: Agreed

File: incremental-font-transfer/src/font_patch.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (garretrieger) [reply incremental-font-transfer/src/font_patch.rs]: Alright, I've reworked the API and hidden both compatibility ID and encoding format.

File: incremental-font-transfer/src/font_patch.rs [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+}
+
+impl IncrementalFontPatchBase for FontRef<'_> {
+    fn apply_patch(
+        &self,
+        compatibility_id: &[u32; 4],
+        patch_data: &[u8],
+        encoding: PatchEncoding,
+    ) -> Result<Vec<u8>, ReadError> {
+        // TODO(garretrieger): Return a custom error type instead of ReadError?
```
[the remark this answers is not in the document]
Comment (garretrieger) [reply incremental-font-transfer/src/font_patch.rs]: Added a new error type.

File: incremental-font-transfer/src/font_patch.rs [hunk 5]
```diff
@@ -67,107 +91,88 @@ impl IncrementalFontPatchBase for FontRef<'_> {
...
-
-            if stream_length as usize > table_patch.brotli_stream().len() {
-                return Err(ReadError::OutOfBounds);
-            }
+        let offset = offset.get().to_u32();
+        let next_offset = next_offset.get().to_u32();
+        let Some(stream_length) = next_offset
+            .checked_sub(offset)
+            .and_then(|v| v.checked_sub(STREAM_START))
+        // brotli stream starts at the (u32 tag + u8 flags + u32 length) = 9th byte
```
Comment (cmyr) [inline incremental-font-transfer/src/font_patch.rs:173]: this comment is now in the wrong place I think?

**Review by cmyr — APPROVED** [review]
👍

**Pull request merged**